src/Pate/AssumptionSet.hs

```diff
@@ -0,0 +1,340 @@
+  show asms = show (PP.pretty asms)
+
+instance OrdF (W4.SymExpr sym) => Monoid (AssumptionSet sym) where
+  mempty = AssumptionSet mempty MapF.empty
+
+mergeExprSetFMap ::
+  OrdF (W4.SymExpr sym) =>
+  OrdF a =>
+  Proxy sym ->
+  MapF.MapF (W4.SymExpr sym) (SetF a) ->
+  MapF.MapF (W4.SymExpr sym) (SetF a) ->
+  MapF.MapF (W4.SymExpr sym) (SetF a)
+mergeExprSetFMap _sym map1 map2 =
+  MapF.mergeWithKey
+    (\_ eset1 eset2 -> Just (eset1 <> eset2))
+    id
+    id
+    map1
+    map2
+
+
+ppBinds ::
+  W4.IsExpr (W4.SymExpr sym) =>
+  Proxy sym ->
+  MapF.MapF (W4.SymExpr sym) (ExprSet sym) ->
+  PP.Doc a
+ppBinds sym bnds =
+  let bs = [ W4.printSymExpr e <+> "-->" <+>  ppExprSet sym es | MapF.Pair e es <- MapF.toList bnds ]
+  in PP.sep (zipWith (<+>) ("[" : repeat ",") bs) <+> "]"
+
+
+-- | Lift an expression binding environment into an assumption set
+fromExprBindings ::
+  forall sym.
+  W4.IsSymExprBuilder sym =>
+  OrdF (W4.SymExpr sym) =>
+  ExprBindings sym ->
+  AssumptionSet sym
+fromExprBindings binds = AssumptionSet { asmPreds = mempty, asmBinds = MapF.map SetF.singleton binds }
+
+
+fromPred ::
+  forall sym.
+  W4.IsSymExprBuilder sym =>
+  W4.Pred sym ->
+  AssumptionSet sym
+fromPred p = AssumptionSet (SetF.singleton p) MapF.empty
+
+exprBinding ::
+  forall sym tp.
+  W4.IsSymExprBuilder sym =>
+  -- | source expression
+  W4.SymExpr sym tp ->
+  -- | target expression
+  W4.SymExpr sym tp ->
+  AssumptionSet sym
+exprBinding eSrc eTgt = case testEquality eSrc eTgt of
+  Just Refl -> mempty
+  _ -> mempty { asmBinds = (MapF.singleton eSrc (SetF.singleton eTgt)) }
+
+natBinding ::
+  forall sym.
+  W4.IsSymExprBuilder sym =>
+  -- | source expression
+  W4.SymNat sym ->
+  -- | target expression
+  W4.SymNat sym ->
+  AssumptionSet sym
+natBinding n1 n2 = exprBinding (W4.natToIntegerPure n1) (W4.natToIntegerPure n2)
+
+ptrBinding ::
+  W4.IsSymExprBuilder sym =>
+  CLM.LLVMPtr sym w ->
+  CLM.LLVMPtr sym w ->
+  AssumptionSet sym
+ptrBinding (CLM.LLVMPointer reg1 off1) (CLM.LLVMPointer reg2 off2) =
+  let
+    regBind = natBinding reg1 reg2
+    offBind = exprBinding off1 off2
+  in (regBind <> offBind)
+
+
+
+macawRegBinding ::
+  W4.IsSymExprBuilder sym =>
+  MS.ToCrucibleType tp ~ MS.ToCrucibleType tp' =>
+  sym ->
+  -- | value to rebind
+  PSR.MacawRegEntry sym tp ->
+  -- | new value
+  PSR.MacawRegEntry sym tp' ->
+  AssumptionSet sym
+macawRegBinding _sym var val = do
+  case PSR.macawRegRepr var of
+    CLM.LLVMPointerRepr _ -> ptrBinding (PSR.macawRegValue var) (PSR.macawRegValue val)
+    CT.BoolRepr -> exprBinding (PSR.macawRegValue var) (PSR.macawRegValue val)
+    CT.StructRepr Ctx.Empty -> mempty
+    _ -> panic Rewriter "macawRegBinding" ["Unexpected macaw types"
+                                          , show (PSR.macawRegRepr var)
+                                          , show (PSR.macawRegRepr val)
+                                          ]
+
```

Comment:
These would all benefit from saying something about what the meaning of the result is (and hopefully that is in terms of the description of `AssumptionSet`)